### DISCUSSION TITLE
Should Custom Method return non-success code from remote or generate JSON-RPC error?

### DISCUSSION THREAD

**Discussion by bugbytesinc** [WalletConnect/walletconnect-monorepo#4002] [Developer Support]
Looking for guidance on the Zen of how to use Wallet Connect for a custom method.  The method passes thru a transaction from a dApp thru the wallet to the Hedera Network.  The Hedera network returns grpc/http success with a hedera custom status code in the GRPC message body (`precheck code`)  This custom code indicates if the transaction was accepted for processing or rejected for any particular reason, like missing a signature for example.  This `precheck code` does not necessarily indicate that processing of the transaction and inclusion of the transaction in ledger will succeed, only that it passed or failed the initial prerequisites.   

_We are having a debate in the hedera community_ if the Wallet Connect Controller/Wallet should intercept the `precheck code` and generate a lower level JSON-RPC error response when it is non-success (which means identifying an Error `code` and custom `data` structure to convey the information) or just simply return the `precheck code` for the `dApp` to examine and act as appropriate.  It seems both approaches have merits, since same is usually better than better in these cases, I’d like to reach out to the wallet connect community/devs and ask how they see the intent of the Wallet Connect library designs and the ”wallet connect way” of handling this use case?

(apologies for this being similar to https://github.com/orgs/WalletConnect/discussions/3789 what we (hedera community) decide above impacts how deep we need to dive into that question down the road)

**Comment by bumblefudge** [comment]
hey again-- not 100% sure I understand the question but I think you might be confusing a little what is [CAIP-25](https://chainagnostic.org/CAIPs/caip-25) and what is [CAIP-27](https://chainagnostic.org/CAIPs/caip-27).  25 is for establishing the wallet-dapp connection with detailed capability and permissions negotiated and stored in mutual state.; it has its own error codes (which are kind of limited to err on the side of opacity to avoid the "browser fingerprinting" pattern that is central to browser deanonymization in web1/2). 27 is a general-purpose RPC call envelope, which basically just forwards an RPC call without needing to understand it and brings back a respnse without needing to understand it-- a "dumb pipe" as they say.  AFAIK there isn't any chain-specific 

I assume using CAIP-27 would be the more appropriate way to make the dApp responsible for the error codes and success semantics, right?  Documenting these in the CAIP-27 profile for the Hedera namespace might be a helpful pointer to the appropriate HIPs and docs that a Hedera newb would need to support Hedera wallets in a multichain wallet.  AFAIK wallet connect qua relay network does not _itself_ handle chain/node errors at all, just route them across wallets and routers; i don't know if catching them, parsing them, and automatically resubmitting a slightly-modified version of the initial txn is in scope for the network or any current add-on products, but i could be wrong.

If, conversely, you were actually talking about error codes or other Hedera-specific details about the _initial wallet connection_, and not transaction/node error codes, then maybe a CAIP-25 profile for implementers and/or a PR into the universal-provider SDK with namespace-specific parsing/process is what's needed? the 25/27 division of labor is, ideally, also the wallet connect libraries/dapp division of labor, as i understand it.

In any case, feel free to "tag me in" anywhere in the multichain or Hedera-specific repos where you guys are having this debate (especially in HIPs, as I'm always game to familiarize myself with new IP processes!) and I'll try to give more specific advice as the debate comes into focus.

**Reply by bugbytesinc** [reply]
Thanks for responding, yes there is definitely some confusion, debate on the best way forward with a Wallet Connect friendly way.  The HIP in question is here: https://hips.hedera.com/hip/hip-820 

A good case study for this is the `hedera_sendTransactionOnly` custom method (https://hips.hedera.com/hip/hip-820#hedera_sendtransactiononly) The crux of the problem comes down to the value returned in the `precheckCode`.  This value is the status code returned from the Hedera gRPC endpoint, if it is zero, the TX was accepted for further processing (does not guarantee success), if it is non zero, it is one of many predefined error codes explaining why it was rejected.  The gRPC endpoint returns http 200 regardless of this value.  Therefore, I thought it would be best to pass that value as the return value on the “happy path” and return it, letting the dApp examine the code for error.  That is the way the HIP is currently proposed.  

There is a counter-proposal in discussion to change this workflow.  If the gRPC endpoint returns a non-zero precheck code, this alternative approach wants the wallet to generate a lower level response at the jsonrpc level.  Where I get confused is, does this break out of the intended model for custom session methods?  How would this be implemented with the current wallet connect libraries?  By simply throwing an exception in the handler?

What they are proposing is returning an JSON-RPC error object down the wire, specifically siting the error object from https://www.jsonrpc.org/specification (Section 5.1 Error object)

There is a line in CAIP-27:

> Response
>
> Upon succesful validation, the respondent will submit or route the request to the targeted network. If the targeted network returns a response to the respondent, the respondent MAY forward this response to the caller. Constraints on, metadata about, or envelopes for response-forwarding MAY be set by namespace profiles of this CAIP.
>
> Similarly, error messages depend on the design of a given namespace, and MAY be defined by a namespace profile of this CAIP.

I believe the guidance therein lies above, but I am not having clarity on how to interpret it in our particular case.

Some of the discussion can be found here https://github.com/hashgraph/hedera-improvement-proposal/discussions/819

Much appreciate your feedback!
